### PR TITLE
[REF] Refactor assets bundles ormcache

### PR DESCRIPTION
**Pull request by Xavier-Do**
## Warning
If changing the ormcache keys and placement was the main goal of this pr, a `self.env['ir.qweb'].clear_caches()` was removed
This is maybe the change that can have the must impact and may create unexpected bug. It was replaced by more precise actions. See the corresponding **Remove cache invalidation** section for more info.

## Introduction

The current ormcache for assets bundle are not always _ideal_.
- Some cache key can have multiple values leading to the same result (example, lang defines rtl)
- Some cache key are useless (too many context key related to templates logic)
- Some cache key are not totally enough and works by "chance"
- Part of the computation stored in the ormcache is actually fast to generate (node attributes)

### Move cache from _get_asset_content to _get_asset_paths

The `_get_asset_content` cache has many cache key that are related to a post-processing of the `_get_asset_paths` result. (node attributes)
`_get_asset_paths` can be slow because of file system access, but the post processing is actually just filtering files, generating node attributes for external assets, .... 
Moving the cache to _get_asset_content will have the benefit to create less duplicates entries in the ormcache as well as less cache miss.

To simplify the `css` and `js` parameters are removed since they _only_ filter the output of get_paths, the heavy part of globing the file will be done before that. Anyway, they are both `True` when called from `_get_asset_content`, and the only other call, in `_get_related_bundle` don't really need to filter them since it is not a critical part regarding performance, and the result will remain the same.

### Better `_get_asset_paths`  cache key

The initial orm cache key was using `_get_template_cache_keys`, a little overkill and creating duplicates entries again. The only context key needed is website_id for `_get_related_assets`.

The context key is actually not enough, the website_id depends on more parameters than that:
- request.session.get('force_website_id')
- request.httprequest.host
- existing websites

The idea here is to call `get_current_website` instead of using all parameters that could define the website.

In the same spirit of `_get_template_cache_keys` `_get_assets_params` can be overridden to give extra params that are useful to list assets path. Those params are computed before entering the method `_get_asset_paths`. 
The only parameter here is the website_id.

Other parameter like  `defer_load`, `lazy_load`, `media` are not useful since the cache was moved but this will actually change even further in next point.

### Result

Before:
```
@tools.ormcache('bundle', 'defer_load', 'lazy_load', 'media', 'tuple(self.env.context.get(k) for k in self._get_template_cache_keys())')
```
After
```
@tools.conditional(
        'xml' not in tools.config['dev_mode'],
        tools.ormcache('bundle', 'tuple(sorted(assets_params.items()))'),
    )
```
The conditional part is actually to make it consistent with the assets_node, and enforce an expected result to discover new files in dev mode (was working before only by cache invalidation side effect) The final cache key is actually : (`bundle_name`, `website_id`)

## 2. Improve _generate_asset_nodes cache

The main reason of the orm cache is the slowness of the validation of the assets. This includes:
- listing files (_get_asset_content, see previous point)
- **computing version**

The cache key was, like the previous point, depending on
- `debug`
The only relevant value for debug is _contains assets_
We don't need to make the difference between debug='', debug='1', debug='test', and 'debug=assets', 'debug=tests,assets', ...
- `defer_load`, `lazy_load`, 
Those values are only useful to generate html node, a leight-weight operation that does not really needs to be in cache. 
The main idea to remove them from the ormcache key is simply to generate the nodes outside the ormcached values.

- `media`
The idea is the same as  `defer_load` and `lazy_load`, but `media` was also used in the generation and it looks useless if we have the media on the node. Since media is not used to generate the attachment url, it doesn't make sence to use it in the generation. Also, no t-call-asset with a media= was found in the master code base.
- `async_load`
This one is similar to `defer_load` and `lazy_load` but it looks like it wasn't used anymore. This was simply removed.
- context.get('lang')
The only information needed is the direction, rtl or ltr. This means en and fr languages, despite sharing the same css assets, will duplicate the ormcache entries.
- `_get_template_cache_keys`
Only the lang and webiste where really relevant in this flow. Other keys are actually useless in this flow.

As explained in the previous point, some information used in the generation where not in the orm cache key
- `self.env.user.lang` if there is no lang in the context
- `request.session.get('force_website_id')`
- `request.httprequest.host`
- ...

The proposed solution is to:
- extract any information needed from the context, request, environment before entering the ormcache, reduce it to the minimal possible set of values needed
```
    rtl = self.env['res.lang']._lang_get_direction(self.env.context.get('lang') or self.env.user.lang) == 'rtl'
    assets_params = self.env['ir.asset']._get_assets_params()  # website_id
    debug_assets = debug and 'assets' in debug
```

and remove a leightweight part of the logic

```
    def _get_asset_nodes(self, bundle, css=True, js=True, debug=False, defer_load=False, lazy_load=False, media=None):
        links = self._get_asset_links(bundle, css=css, js=js, debug=debug)
        return self._links_to_nodes(links, defer_load=defer_load, lazy_load=lazy_load, media=media)
```

Where _get_asset_links is the heavy cached part, and _links_to_nodes is the lightweight part generating the nodes based on the `defer_load`, ....

Additional notes:
- data-asset-version and data-asset-bundle are removed from the node since they don't seem to be used anymore since 65d70acdbfb1027f0937b5af7c4f3e9a571a4c6e
- async_load is removed since there is no occurence of this in the code.
- a small hack is still needed to pass javascript content instead of links, this is only to manage css compile error and will hopefully be removed in the future.
- a context key is still in use to generate the bundle, the `commit_assetsbundle` but it has no impact on content and will hopefully be removed in the future.

## Add test for ormcache hit/miss

In this context, hit/miss is about having the same cache key for the same result. This test demonstrates the current state, were entries are create in the ormcache only if the key is really different and will lead to a different result.

## Remove cache invalidation

This cache invalidation is quite aggressive since every time an asset bundle is updated, all workers will clear their cache.

The concerned cache by this clear_cache is `_generate_asset_nodes_cache` trough `_get_asset_nodes`.

The cache is ignored, both in dev=xml and debug=assets.

This clear cache was made conditional in https://github.com/odoo/odoo/commit/553ea82f8135454b90888bf200372744729d7b20 but this does not solve an issue we can have in production.

Lets imagine a clean starting state
- all sources are updated
- all workers are restarted.

The orm caches are all empty, but since the sources changed, all bundles will be recomputed. This means that every bundle updated in database with save_attachement will invalidate the cache of all workers. Rendering a pdf report of any kind using a specific bundle will invalidate all cache. Starting a debug=assets for the first time will invalidate all cache, even if the cache is not used in this case.

But for a regenerated bundle we would expect the ormcache to be:
- empty (did not generate the same bundle yet)
- have the same value (concurrent generation of the same bundle)

Having a different value would mean that the bundle was generated with another version of the sources. In this case it is maybe even better not to invalidate the cache since it could lead to an invalidation war between two workers.

The only case where invalidating this cache is useful is when a bundle changes, Usually if an ir_asset is created, modified, ...

There is still another rare but possible possibility to have a 404 if the transaction is roll backed after populating the assets node cache.
In this case, we only need to clear the cache locally in case of rollback.

## Copy bundle when possible (2nd commit)

If get_attachments fails, the next step will be to generate the attachments from scratch, a slow operations.

When creating a new website, all assets bundle would actually be similar to their version without website, but the url is different.

This can be visible because the first loading of /web is slow after creating a new website: the website_id is forced in the session and the `assets_backend` are regenerated, identical to the original ones.

This commit proposes to try to find an attachments with different extra but the same uniquifier when possible and copy it's content.

Note that just returning the other attachment url may work, but it would be confusing to randomly have links to assets coming from another website_id. This would also be a problem if the original attachment is unlinked, forcing to recompute it for other websites.

Good to now, since the content is the same, no duplication of the content should appear in the file store, just an entry in the database.

## Third commit

Because of all previous changes, we can simplify even futher. The main idea is to avoid to be forced to call `_get_asset_content` before  `_get_asset_bundle` (or `AssetsBundle`). We can get rid of `AssetsBundleMultiWebsite` by making the small change of behaviour between the two work by calling an `ir.asset` method `_get_asset_extra` to generate the extra parameter. This in fine allows to have less overrides. 
The "remains" (external assets url) of get_assets_content are also stored on the assets_bundle.

### CODE REVIEW

**Comment by Xavier-Do** [comment]
The test `test_05_specific_website_editor` is failing and after some investigation it is linked to a bug in the test were the page is `/undefined` leading to a 404 and the fronted assets are not committed for website 1. 

It took some time to figure out but  the reason is wasn't failing before is because of the pregenerate of assets in website, copying attachment for website 1. 

With the new version (last commit), the assets are not copied in pregenerate anymore but on demand. Since the demand is done in a 404 page, the transaction is not committed and the asset bundle is missing. This is actually a known bug.

This test will be fixed in another pr.

File: odoo/addons/base/models/ir_qweb.py [hunk 1]
```diff
@@ -2154,7 +2154,7 @@ def _compile_directive_call_assets(self, el, compile_context, level):
         xmlid = el.attrib.pop('t-call-assets')
         css = self._compile_bool(el.attrib.pop('t-css', True))
         js = self._compile_bool(el.attrib.pop('t-js', True))
-        async_load = self._compile_bool(el.attrib.pop('async_load', False))
+        # async_load support was removed
```
Comment (VincentSchippefilt) [inline odoo/addons/base/models/ir_qweb.py:2157]: should be in a separate commit with this in the message

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 2]
```diff
@@ -1925,3 +1891,137 @@ def test_37_path_can_be_an_attachment(self):
...
+        asset_keys = [key for key in keys if key[0] == 'ir.asset' and '_get_asset_paths' in str(key[1])] # ignore topological sort entry
+        qweb_keys = [key for key in keys if key[0] == 'ir.qweb']
+        return asset_keys, qweb_keys
+
+    def test_assets_node_orm_cache_usage_debug(self):
+        self.env.registry._clear_cache()
+
+        asset_keys, qweb_keys = self.cache_keys()
+        self.assertEqual(len(asset_keys), 0)
+        self.assertEqual(len(qweb_keys), 0)
```
Comment (VincentSchippefilt) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py:1908]: you can refactor this function and all the tests to
`assert_cache_keys(expected_number_assets_keys, expected_number_qweb_keys)`  --> you will remove 2 lines of copy/pasted call for each assert

**Comment by VincentSchippefilt** [comment]
in commit message of second commit: 
Good to ~~now~~ know, since the content is the same, no duplication of the
content should appear in the filestore, just an entry in the database.

File: addons/website/tests/test_ui.py [hunk 3]
```diff
@@ -222,7 +222,10 @@ def test_04_website_navbar_menu(self):
         self.start_tour("/", 'website_navbar_menu')
 
     def test_05_specific_website_editor(self):
+        asset_bundle_xmlid = 'website.assets_wysiwyg'
         website_default = self.env['website'].search([], limit=1)
+
+
```
Comment (rco-odoo) [inline addons/website/tests/test_ui.py:228]: one empty line is enough

File: odoo/addons/base/models/assetsbundle.py [hunk 4]
```diff
@@ -265,25 +237,76 @@ def get_attachments(self, extension, ignore_version=False):
...
+                    attachment_id=attachment.id,
+                    unique=unique,
+                    extra=extra,
+                    name=self.name,
+                    extension=extension,
+                )
+                values = {
+                    'url': url,
+                }
+                attachment.write(values)
```
Comment (rco-odoo) [inline odoo/addons/base/models/assetsbundle.py:296]:
```suggestion
                attachment.url = url
```

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py [hunk 5]
```diff
@@ -1925,3 +1890,137 @@ def test_37_path_can_be_an_attachment(self):
...
+        self.assertEqual(len(asset_keys), 1)
+        self.assertEqual(len(qweb_keys), 1)
+
+        # in debug=assets, the ormcache is not used for _generate_asset_links_cache
+        self.env['ir.qweb']._get_asset_nodes('web.assets_backend', debug='assets')
+        asset_keys, qweb_keys = self.cache_keys()
+        self.assertEqual(len(asset_keys), 1)
+        self.assertEqual(len(qweb_keys), 1)
+
+
```
Comment (rco-odoo) [inline odoo/addons/test_assetsbundle/tests/test_assetsbundle.py:1931]: style nitpicking: methods are separated by a single blank line
```suggestion

```

File: odoo/addons/test_assetsbundle/tests/test_assetsbundle.py (hunk 2 above)
Comment (Julien00859) [reply odoo/addons/test_assetsbundle/tests/test_assetsbundle.py:1908]: would seem over-engineered to me, the current way is very readable

File: addons/website/models/ir_asset.py [hunk 6]
```diff
@@ -10,35 +10,48 @@ class IrAsset(models.Model):
...
-    def _get_related_assets(self, domain):
-        website = self.env['website'].get_current_website(fallback=False)
-        if website:
-            domain += website.website_domain()
-        assets = super()._get_related_assets(domain)
-        return assets.filter_duplicate()
-
-    def _get_active_addons_list(self):
+    def _get_assets_params(self):
+        params = super()._get_assets_params()
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply addons/website/models/ir_asset.py:14]: btw, this method is a nice simplification for website :+1:

**Review by Julien00859 — COMMENTED** [review]
Very very early review, I only started to scratch the surface, I basically just read those two functions. Btw moving the ormcache from content to path is really the plus-value, it shouldn't had been added to content in the first place :(

File: odoo/addons/base/models/ir_asset.py [hunk 7]
```diff
@@ -91,7 +90,25 @@ def unlink(self):
     active = fields.Boolean(string='active', default=True)
     sequence = fields.Integer(string="Sequence", default=DEFAULT_SEQUENCE, required=True)
 
-    def _get_asset_paths(self, bundle, css=False, js=False):
+    def _get_assets_params(self):
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply odoo/addons/base/models/ir_asset.py]: done

File: odoo/addons/base/models/ir_qweb.py [hunk 8]
```diff
@@ -2472,80 +2484,91 @@ def _get_cached_values(self, cache_key, get_value):
...
+    def _generate_asset_links_cache(self, bundle, css=True, js=True, assets_params=None, rtl=False):
+        return self._generate_asset_links(bundle, css, js, False, assets_params, rtl)
 
-    @tools.ormcache('bundle', 'defer_load', 'lazy_load', 'media', 'tuple(self.env.context.get(k) for k in self._get_template_cache_keys())')
-    def _get_asset_content(self, bundle, defer_load=False, lazy_load=False, media=None):
-        asset_paths = self.env['ir.asset']._get_asset_paths(bundle=bundle, css=True, js=True)
+    def _get_asset_content(self, bundle, assets_params=None):
+        if assets_params is None:
+            assets_params = self.env['ir.asset']._get_assets_params()  # website_id
+        asset_paths = self.env['ir.asset']._get_asset_paths(bundle=bundle, assets_params=assets_params)
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply odoo/addons/base/models/ir_qweb.py]: _get_asset_content is used somewhere else. Could be modified but this can be done in another pr.

File: addons/website/models/ir_asset.py [hunk 9]
```diff
@@ -10,35 +10,48 @@ class IrAsset(models.Model):
...
+    def _get_asset_extra(self, extra, website_id=None, **params):
+        extra = super()._get_asset_extra(extra, **params)
+        if extra == '%':
+            return extra
+        website_id_path = website_id and ('%s/' % website_id) or ''
+        return website_id_path + extra
+
+    def _get_related_assets(self, domain, website_id=None, **params):
+        if website_id:
+            domain += [('website_id', 'in', (False, website_id))]
```
Comment (rdeodoo) [inline addons/website/models/ir_asset.py:27]:
```suggestion
            domain += self.env['website'].website_domain(website_id)
```

File: odoo/addons/base/models/ir_qweb.py [hunk 10]
```diff
@@ -2472,80 +2484,91 @@ def _get_cached_values(self, cache_key, get_value):
...
-        files, remains = self._get_asset_content(bundle, defer_load=defer_load, lazy_load=lazy_load, media=css and media or None)
-        asset = self._get_asset_bundle(bundle, files, env=self.env, css=css, js=js, debug=debug)
-        remains = [node for node in remains if (css and node[0] == 'link') or (js and node[0] == 'script')]
-        return remains + asset.to_node(css=css, js=js, async_load=async_load, defer_load=defer_load, lazy_load=lazy_load)
+                external_asset.append((path, None))
+        return (files, external_asset)
+
+    def _get_asset_bundle(self, bundle_name, css=True, js=True, debug_assets=False, rtl=False, assets_params=None):
+        if assets_params is None:
+            assets_params = self.env['ir.asset']._get_assets_params()
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply odoo/addons/base/models/ir_qweb.py]: The one in get_asset_content is a fallback in case it was not called by _get_asset_bundle. And it is given to the bundle so it is needed here.

File: addons/website/models/ir_asset.py [hunk 11]
```diff
@@ -10,35 +10,48 @@ class IrAsset(models.Model):
...
+        if extra == '%':
+            return extra
+        website_id_path = website_id and ('%s/' % website_id) or ''
+        return website_id_path + extra
+
+    def _get_related_assets(self, domain, website_id=None, **params):
+        if website_id:
+            domain += self.env['website'].website_domain(website_id)
+        assets = super()._get_related_assets(domain, **params)
+        return assets.filter_duplicate(website_id)
```
Comment (rdeodoo) [inline addons/website/models/ir_asset.py:29]: As the base implementation is the following:
```py
    def _get_related_assets(self, domain):
        """
        Returns a set of assets matching the domain, regardless of their
        active state. This method can be overridden to filter the results.
        :param domain: search domain
        :returns: ir.asset recordset
        """
        # active_test is needed to disable some assets through filter_duplicate for website
        # they will be filtered on active afterward
        return self.with_context(active_test=False).sudo().search(domain, order='sequence, id')
```

Wouldn't it be better to add `**kw` on this base implementation and keep the same signature here, poping `website_id` from it before calling `super()`?

The different signature between those override seems weird to me, especially once other modules will / might start writing their own different one too.

Or is it actually a good practice / pattern the way it is currently implemented?

Comment (Xavier-Do) [reply addons/website/models/ir_asset.py:29]: It is subjective I think I was a little `partagé` between the two solutions but my linter indicates when I have a unused parameter but not when the ovverrides have different signatures. And I actually like this pattern. But juc will answer in details why.

Comment (Julien00859) [reply addons/website/models/ir_asset.py:29]: > The different signature between those override seems weird to me

It is normal you find it weird, this is a way to implement method *overload* (=adding a new signature to a method in a child class) in Python. There are two ways to do overload, (1) change the type of one of the arguments (e.g. list vs tuple), (2) add new arguments. (1) is much more common than (2) in the wild and it is solved generally solved in Python thanks to duck-typing (usually [datamodel](https://docs.python.org/3/reference/datamodel.html) + [collection.abc](https://docs.python.org/3/library/collections.abc.html)) meaning we usually don't need to do any Java-like type-1 overload. Here we wanted a type-2 overload, those are much less common, specially in Python that as no dedicated syntax for it unlike Java, what you are reading here is a way to implement type-2 overload.

> might start writing their own different one too

This is actually part of the pattern, we *enable* method overload on that method, other classes can overload it their own way and add their new parameters. This is expected.

> Or is it actually a good practice

This is not good/bad practice, this is one way of implementing method overload in Python. Are there other solutions than method overload to solve this problem: surely, would they be better: maybe, is the current solution good/simple enough: we think so

> Wouldn't it be better to add **kw on this base implementation and keep the same signature here, poping website_id from it before calling super()?

This would work too and we would had implemented it this way:

```py
def _get_related_assets(self, domain, **kw: dict):
    webside_id = kw.pop('webside_id', None)
    super()._get_related_assets(domain, **kw)
    ...

def _get_related_assets(self, domain, **kw: dict):
    if kw:
        raise TypeError(f"The base implementation takes no extra keyword arguments, but some are present. Thoses KW should had been consumed by overrides: {kw}")
    ...
```

The benefit of using the current implementation using method overload is that all the extra kwargs that will be used within the method are explicit on the function signature. It is also explicit that the base implementation takes no extra kwargs and the `raise TypeError(...)` is automatically done by Python.

Depending on whether an override/overload will/will not *consume* (use it and remove it from the kwargs),

Override:
```py
def foo(**kwargs):
    x = kwargs.get("x", None)  # does not consume
    y = kwargs.pop("y", None)  # does consume
    super().foo(**kwargs)
```

Overload:
```py
def foo(x, y, **kwargs):
    super().foo(x, **kwargs)
```

---

If you do not like this pattern and rather we implement a more classic override (like in the Override example) it is your call, we basically implemented it only for website :p

Comment (rdeodoo) [reply addons/website/models/ir_asset.py:29]: > This would work too and we would had implemented it this way:

That's what I had in mind/suggested in my previous message indeed, except for the "raise no kw", is that really needed?

> If you do not like this pattern and rather we implement a more classic override (like in the Override example) it is your call, we basically implemented it only for website :p

No it's fine, it's not usual, at least in Odoo and bugged me when reading it, also the part where you need to remove params when calling the super, not that straightforward.
So I was more curious than anything else.
Thanks for the detailed explanation, don't need to change the implementation :+1:

Comment (Julien00859) [reply addons/website/models/ir_asset.py:29]: > you need to remove params when calling the super, not that straightforward

yes you are right :+1:, this is uncommon to have a super() call that is different from the function signature

Comment (Xavier-Do) [reply addons/website/models/ir_asset.py:29]: As juc said since it is mainly for website I have no issue to change that (but as you said without the raise). It is just a pattern I like but there is no other valid rational reason to use it except maybe for the current linting rules that are in place.

Comment (Xavier-Do) [reply addons/website/models/ir_asset.py:29]: I made a quick branch to show the difference with another pattern:

https://github.com/odoo-dev/odoo/compare/master-refactor-assets-ormcache-xdo..master-example-pattern-xdo?expand=1

One drawback of this solution is that you may need to modify all overrides if you want to add any param to the _get_asset_params method.

We can imagine that it will only have an effect on _get_related_assets meaning that the
_get_active_addons_list would be useless and you will need to do it. But actually, the _get_asset_extra would need to me modified because it is not correct to change the behavior of generation of assets based on a new parameter without adding this info in the extra so it looks like a good idea to force this override.

Anyway, I don't think anyone else than website will override this method. The whole _get_asset_params is maybe overkill since we only need the website_id. But base is not aware of that unfortunately. I think this implementation is a good compromise between extensibility and readability, even if how it works can look a little complicated, almost wrong from the outside when looking closely.

Comment (rdeodoo) [reply addons/website/models/ir_asset.py:29]: > I made a quick branch to show the difference with another pattern:
>
> https://github.com/odoo-dev/odoo/compare/master-refactor-assets-ormcache-xdo..master-example-pattern-xdo?expand=1

Note that I would've gotten rid of the `website_id` in the signature and have it inside the `**kw` but as you said, don't worry about it, I got the explanation I am fine with it :)

**Comment by Julien00859** [comment]
@robodoo r+

**Comment by Xavier-Do** [comment]
@robodoo rebase-ff

**Comment by Xavier-Do** [comment]
@robodoo r+
